src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs

```diff
@@ -270,6 +252,69 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
             return selectedFormatter;
         }
 
+        private IList<MediaTypeSegmentWithQuality> GetMediaTypes(
+            IList<StringSegment> contentTypes,
+            HttpRequest request)
+        {
+            // Will fix the list with object pooling
+            var result = new List<MediaTypeSegmentWithQuality>();
+            AcceptHeaderParser.ParseAcceptHeader(request.Headers[HeaderNames.Accept], result);
+            for (int i = 0; i < result.Count; i++)
+            {
+                if (!RespectBrowserAcceptHeader &&
+                    IsAcceptAllMediaTypesMediaType(result[i]))
+                {
+                    result.Clear();
+                    return result;
+                }
+
+                RemoveMediaTypeFromResultsIfNotAcceptable(result, i, contentTypes);
+            }
+
+            SortMediaTypesByQuality(result);
+
+            return result;
+        }
+
+        private static bool IsAcceptAllMediaTypesMediaType(MediaTypeSegmentWithQuality mediaTypeWithQuality)
+        {
+            return mediaTypeWithQuality.MediaType.StartsWith("*/*", StringComparison.OrdinalIgnoreCase);
+        }
+
+        private static void RemoveMediaTypeFromResultsIfNotAcceptable(
+            IList<MediaTypeSegmentWithQuality> mediaTypes,
+            int index,
+            IList<StringSegment> acceptableMediaTypes)
+        {
+            if (!InAcceptableMediaTypes(mediaTypes[index].MediaType, acceptableMediaTypes))
+            {
+                mediaTypes.RemoveAt(index);
+            }
+        }
+
+        private static void SortMediaTypesByQuality(List<MediaTypeSegmentWithQuality> result)
+        {
+            result.Sort((left, right) => left.Quality > right.Quality ? 1 : (left.Quality == right.Quality ? 0 : -1));
+        }
+
+        private static bool InAcceptableMediaTypes(StringSegment mediaType, IList<StringSegment> acceptableMediaTypes)
+        {
+            if (acceptableMediaTypes.Count == 0)
+            {
+                return true;
+            }
+
+            for (int i = 0; i < acceptableMediaTypes.Count; i++)
+            {
+                if (MediaTypeComparisons.IsSubsetOf(mediaType, acceptableMediaTypes[i]))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
```

Comment:
This is way less clear with all these little helper methods that have 2 statements